File: 2024/08_Resonant_Collinearity/Day08.kt
// Day 8: Resonant Collinearity
// https://adventofcode.com/2024/day/8

import java.io.File

fun readInput(filename: String): String {
    return File("$filename.txt").readText()
}

fun parsedGrid(input: String): List<MutableList<Char>> =
    input.lines()
        .filter { it.isNotBlank() }
        .map { it.toMutableList() }

fun isInBounds(grid: List<List<Char>>, row: Int, col: Int): Boolean =
    row in grid.indices && col in grid[0].indices

fun computeAntennaNodes(grid: List<List<Char>>): Map<Char, List<Pair<Int, Int>>> =
    grid.flatMapIndexed { row, line ->
        line.mapIndexedNotNull { col, cell ->
            if (cell != '.') cell to Pair(row, col) else null
        }
    }.groupBy({ it.first }, { it.second })

fun computeAntiNodes(
    grid: List<MutableList<Char>>,
    includeSelfNode: Boolean = false,
    infiniteFrequency: Boolean = false
): Set<Pair<Int, Int>> {
    val antennaMap = computeAntennaNodes(grid)
    val antiNodeSet = mutableSetOf<Pair<Int, Int>>()

    antennaMap.values.forEach { nodes ->
        nodes.forEachIndexed { i, (row1, col1) ->
            for (j in i + 1 until nodes.size) {
                val (row2, col2) = nodes[j]

                if (includeSelfNode) {
                    antiNodeSet += Pair(row1, col1)
                    antiNodeSet += Pair(row2, col2)
                }

                val diffRow = row1 - row2
                val diffCol = col1 - col2

                generateAntiNodes(grid, row2, col2, diffRow, diffCol, infiniteFrequency, antiNodeSet)
                generateAntiNodes(grid, row1, col1, -diffRow, -diffCol, infiniteFrequency, antiNodeSet)
            }
        }
    }
    return antiNodeSet
}

fun generateAntiNodes(
    grid: List<MutableList<Char>>,
    startRow: Int,
    startCol: Int,
    diffRow: Int,
    diffCol: Int,
    infiniteFrequency: Boolean,
    antiNodeSet: MutableSet<Pair<Int, Int>>
) {
    var distanceFactor = 2
    while (true) {
        val newRow = startRow + diffRow * distanceFactor
        val newCol = startCol + diffCol * distanceFactor

        if (!isInBounds(grid, newRow, newCol)) break

        grid[newRow][newCol] = '#'
        antiNodeSet += Pair(newRow, newCol)

        if (!infiniteFrequency) break
        distanceFactor++
    }
}

fun part1(input: String): Int {
    val grid = parsedGrid(input).map { it.toMutableList() }
    val antiNodes = computeAntiNodes(grid, includeSelfNode = false, infiniteFrequency = false)
    grid.forEach { row -> println(row.joinToString("")) }
    return antiNodes.size
}

fun part2(input: String): Int {
    val grid = parsedGrid(input).map { it.toMutableList() }
    val antiNodes = computeAntiNodes(grid, includeSelfNode = true, infiniteFrequency = true)
    grid.forEach { row -> println(row.joinToString("")) }
    return antiNodes.size
}

fun main() {
    val input = readInput("Day08")
    val inputTest = readInput("Day08_test")

    println("Part 1: ${part1(input)}")
    println("Part 1 Test: ${part1(inputTest)}")

    println("Part 2: ${part2(input)}")
    println("Part 2 Test: ${part2(inputTest)}")
}

File: 2024/08_Resonant_Collinearity/Day08.js
// Day 8: Resonant Collinearity
// https://adventofcode.com/2024/day/8

const fs = require('fs');
function readInput(filename) {
  return fs.readFileSync(`${filename}.txt`, 'utf8');
}

function parsedGrid(input) {
  return input
    .split('\n')
    .filter(Boolean)
    .map((row) => row.split(''));
}

function isInBounds(grid, row, col) {
  return row >= 0 && row < grid.length && col >= 0 && col < grid[0].length;
}

function computeAntennaNodes(grid) {
  const rows = grid.length;
  const cols = grid[0].length;
  const antennaMap = new Map();
  for (let row = 0; row < rows; row++) {
    for (let col = 0; col < cols; col++) {
      const antenna = grid[row][col];
      if (antenna !== '.') {
        if (!antennaMap.has(antenna)) {
          antennaMap.set(antenna, []);
        }
        antennaMap.get(antenna).push({ row, col });
      }
    }
  }
  return antennaMap;
}

function computeAntiNodes(grid, includeSelfNode = false, infiniteFrequency = false) {
  const antennaMap = computeAntennaNodes(grid);
  const antiNodeSet = new Set();
  for (const [, nodes] of antennaMap) {
    for (let i = 0; i < nodes.length; i++) {
      for (let j = i + 1; j < nodes.length; j++) {
        const node1 = nodes[i];
        const node2 = nodes[j];

        if (includeSelfNode) {
          antiNodeSet.add(`${node1.row},${node1.col}`);
          antiNodeSet.add(`${node2.row},${node2.col}`);
        }

        const diffRow = node1.row - node2.row;
        const diffCol = node1.col - node2.col;

        generateAntiNodes(grid, node2, diffRow, diffCol, infiniteFrequency, antiNodeSet);
        generateAntiNodes(grid, node1, -diffRow, -diffCol, infiniteFrequency, antiNodeSet);
      }
    }
  }
  return antiNodeSet;
}

function generateAntiNodes(grid, startNode, diffRow, diffCol, infiniteFrequency, antiNodeSet) {
  let distanceFactor = 2;
  while (true) {
    const row = startNode.row + diffRow * distanceFactor;
    const col = startNode.col + diffCol * distanceFactor;

    if (!isInBounds(grid, row, col)) break;

    grid[row][col] = '#';
    antiNodeSet.add(`${row},${col}`);

    if (!infiniteFrequency) break;

    distanceFactor++;
  }
}

function part1(input) {
  const grid = parsedGrid(input);
  const antiNodes = computeAntiNodes(grid, false, false);
  grid.forEach((row) => console.log(row.join('')));
  return antiNodes.size;
}

function part2(input) {
  const grid = parsedGrid(input);
  const antiNodes = computeAntiNodes(grid, true, true);
  grid.forEach((row) => console.log(row.join('')));
  return antiNodes.size;
}

console.log(part1(readInput('Day08'))); // 14
console.time('Part 1');
console.log(part1(readInput('Day08_test'))); // 354
console.timeEnd('Part 1');

console.log(part2(readInput('Day08'))); // 34
console.time('Part 2');
console.log(part2(readInput('Day08_test'))); // 1263
console.timeEnd('Part 2');

/*
.....#..#...#..#..#...##.###.####.#.......#..#....
.#####..##.##..##.#.#.#.....##.#...#.#.#....#.#...
..#.#...###.#.....#.##...######..####C.#...#...#..
#.##.....###.#.##..###....##.#.####.###...##.##..#
.##...#.###.#b.#####.u.#.##.###w#......#.#.3...##.
.##......#.#.##..######.##.#........#...##.#.#.##.
#####..#...#K###.#..#..####....#y#.....#..##..#.##
....#.......#..##.######.#.#.#..#.P5..##.#..B#...#
#..U.#.....##.###.#.2##.#.###.#...#..###.#C..#....
#.....#...###.#######P.##.....w##.#.###C#...#.I#5#
.#.....###..####..###.##..#..##....##.#.x#..#....#
.#.#...###..#....#####...1.......##....#......u#..
#...#.#..###..#..##.#.##.#.####r.####....s.#..#...
##2##.###.#.#...t##T.#.###.#.r..#...###...#..#.I##
#.#....#####..#t##T...#.##.###.###..###....#5.##..
...#.#..#.#.####.#.#.#####..1.#..M.t##.#...####...
.#.V####U..#.##.###...###.#.##..N.S....###...I.#.#
...####.##..#####..###.#.#O.#..v#..#.#....#.S##.#.
.#.####..####.###.#####.s###.....#.#.#...##...3###
..#.6###.###C#..##..##..##Tf#3##.##...#.#..#.#....
#..##.#####F#..#.#.#..V.##.##.#..k#.#N...#..H#..#.
#.##.##.##.###.#..#U#.#.##..##.0#...#.H.y.......##
##.#..######.....#.#..##..#P..###.0.#....#.##...##
##....##.#U#...#.##.#.#R.##..#.#..##s#.....##.##..
###.#####..###.##.####O....##2..#.#.##.H###..B#...
.########..######..######..#...###.#.#.#.S.#.....#
...####.##..c.#...####.#.##...#...##..d....######.
########.###..##F.#.#.##m#..#......#.#.##.#7...#.#
.##.##.###.###..##..#.#..##..#....#.##.S##.##..###
####..j.k.###.#.##.....V#.#..#.##..###....#..#.#..
###.#.#...#.####.L##...K...#.#...#...##..#.#.###..
.#....#F...#..###.##...##..####..#.#M#.###.#...#..
##.....#..#..#..###.#.m1....a#R...#...####...#...#
..#...#4#...#.#####D.#......#.##......##.##.......
k...##.#...#.9##.#.####..#..#.####.###M...#.....#.
#..#.#b.#.#..#......#.##V...h#.###.#....#.###..#..
A##....###9#Xl###.#.#...##.##..R##.........##..##.
4.##....#.#...#..#.##....#D#..######....#.#..#.##.
.A.#.#.#...#..##.#...#..######.M.7##....#...#.##.#
....####..n.#.##..#.#...#.##.###.##.##...#.##.##..
..#j##..b#....#...##.###.p..#######..####.####..#.
...##.#..#.##k..##.#...p.##..#.#####.##.....####.#
.#.#.#.W4..######.##.##X#....###J##.##v#..##.##..#
.##.#...W#d..c.#.###A..#...#.#.###..#.###...#.#..#
.##.#..#.#l##.####..##.#..#o..#.##Y###h.##.......#
#..#.A....#.#.###.###.#.#.##.....##..####.##..#...
#.j..#....##.###.#...#.#.###...#..##..##.######...
a9.lX..##..##...#..#...Y.##...#....#.###....#.###.
..#...#a.###.#..#.#......Y#.##.#.##.#..#..#.##.###
.#.#...#l.#.#.#.###.#.##.###...##...##.###....###.
*/
